Changing envelope definition to  support two fixed modes (#102)
cfrg/draft-irtf-cfrg-opaque: draft-irtf-cfrg-opaque.md

: A unique 32-byte nonce used to protect this Envelope.

ct
: Encoding of encrypted and authenticated `SecretCredentials`.

auth_tag
: Authentication tag protecting the contents of the envelope,

> incorporating `InnerEnvelope` and `CleartextCredentials`.

Reviewer: ```suggestion
covering `InnerEnvelope` and `CleartextCredentials`.
```

The full procedure for constructing `Envelope` and `InnerEnvelope` from

> `SecretCredentials` and `CleartextCredentials` is described in {{finalize-request}}.

Reviewer: ```suggestion
`SecretCredentials` and `CleartextCredentials` is described in {{finalize-request}}.

The `EnvelopeMode` value is specified as part of a configuration. 
```

1. (kU, _) = KeyGen()
2. Z = Evaluate(kU, request.data)
3. Create RegistrationResponse response with
     (Z, pkS)
4. Output (response, kU)
~~~

#### FinalizeRequest {#finalize-request}

~~~

> FinalizeRequest(pwdU, skU, blind, request, response)

Reviewer: Also unrelated, but we can remove `request` here, as it's not used.

Author: It is used in line 468:

```
The full procedure for constructing `Envelope` and `InnerEnvelope` from
`SecretCredentials` and `CleartextCredentials` is described in {{finalize-request}}.
```

Reviewer: I'll address this in a separate PR, since it seems related to the module boundary we discussed.

Steps:
1. N = Unblind(blind, response.data)
2. y = Finalize(pwdU, N, "OPAQUE01")
3. rwdU = HKDF-Extract("rwdU", Harden(y, params))
4. Create secret_credentials with credentials matching those needed to
   construct the `SecretCredentials` structure
5. Create cleartext_credentials with credentials matching those needed to
   construct the `CleartextCredentials` structure

> 6. pt = Serialize(secret_credentials)

Reviewer 2: If SecretCredentials is always equal to the client secret, wouldn't it be more straightforward to put `pt = Serialize(secret)` ?

Author: I think it may be beneficial to keep this level of generality here in case we add support for other modes that set SecretCredentials to something else in the future. This will hopefully also coincide with the changes that will be made in #105 which will pass in a generic credentials parameter.

7. nonce = random(32)
8. pseudorandom_pad = HKDF-Expand(rwdU, concat(nonce, "Pad"), len(pt))
9. auth_key = HKDF-Expand(rwdU, concat(nonce, "AuthKey"), Nh)
10. export_key = HKDF-Expand(rwdU, concat(nonce, "ExportKey"), Nh)
11. ct = xor(pt, pseudorandom_pad)
12. auth_data = Serialize(cleartext_credentials)

> 13. Create InnerEnvelope contents with (nonce, ct)

Reviewer 2: If I'm not mistaken, the InnerEnvelope should also specify the InnerEnvelopeMode, doesn't it?

Author: Good catch, thank you! Added to parameters.

14. t = HMAC(auth_key, concat(contents, auth_data))
15. Create Envelope envU with (contents, t)

> 16. Create RegistrationUpload upload with envelope value (envU, pkU)

Reviewer: Unrelated: should we lift `pkU` to the wrapper protocol, too?

Author: As discussed offline, let's address this API change separately.

~~~

Here, idU and idS are by default set to be equal to the idU and idS supplied as a
credential for the envelope; however, if no such credential was supplied,

> then these values are (indepedently) defaulted to pkU and pkS instead.

Reviewer: ```suggestion
then these values (independently) default to pkU and pkS instead.
```

Future specifications (such as {{I-D.sullivan-tls-opaque}}) MAY introduce other
AKE instantiations.

> The EnvelopeMode value is defined in {{data-types}} and must also be agreed upon

Reviewer: This implies (to me) some sort of negotiation. Since the EnvelopeMode is already in the tuple above, I'd drop that bit:

```suggestion
The EnvelopeMode value is defined in {{data-types}}.
```
